### PR TITLE
[SYCL] Fix deferred buf destruction regression caused by host device removal

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
I'm not familiar with neither this area nor the previous PR... Adding @sergey-semenov

**Comment by KseniyaTikhomirova** [comment]
fixes https://github.com/intel/llvm/issues/14397

File: sycl/source/detail/sycl_mem_obj_t.cpp [hunk 1]
```diff
@@ -205,13 +205,17 @@ void SYCLMemObjT::detachMemoryObject(
...
   bool InteropObjectsUsed =
       !MOwnNativeHandle ||
       (MInteropContext && !MInteropContext->isOwnedByRuntime());
 
-  if (MRecord && MRecord->MCurContext &&
-      MRecord->MCurContext->isOwnedByRuntime() && !InteropObjectsUsed &&
-      (!MHostPtrProvided || MIsInternal)) {
+  if (MRecord &&
+      (!MRecord->MCurContext ||
+       (MRecord->MCurContext && MRecord->MCurContext->isOwnedByRuntime())) &&
```
Comment (sergey-semenov) [inline sycl/source/detail/sycl_mem_obj_t.cpp:217]:
```suggestion
      (!MRecord->MCurContext ||
       MRecord->MCurContext->isOwnedByRuntime()) &&
```
Nit

**Review by sergey-semenov — APPROVED** [review]

File: sycl/source/detail/sycl_mem_obj_t.cpp (hunk 1 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/sycl_mem_obj_t.cpp]: thanks, updated

**Review by sarnex — APPROVED** [review]
test changes LGTM

@KseniyaTikhomirova What is your confidence that this change fixes the cause of the hang? If we don't have very high confidence and re-enable the tests, it could break the CI again.

Thanks.

**Comment by KseniyaTikhomirova** [comment]
> test changes LGTM
> 
> @KseniyaTikhomirova What is your confidence that this change fixes the cause of the hang? If we don't have very high confidence and re-enable the tests, it could break the CI again.
> 
> Thanks.

I was able to reproduce hang on local WIN Gen9 system and with this fix tests passed

**Comment by sarnex** [comment]
Ok, sounds legit to me, thanks.

**Pull request merged**